Iterable Dataset for K2 ASR

pzelasko opened the pull request:
This PR explores how to go forward with #113 and should resolve #115 

I managed to implement a first, working version of the iterable ASR dataset for K2. I also tested it with LibriSpeech train-clean-100 and discovered that I failed to guarantee that `num_frames` field in the manifests always returns correct values (this is mostly related to incosistent rounding behaviour for durations like 16.475). I put some workarounds in the code so that it can collate without issues (a dataloader on train-clean-100 works without issues). I will resolve them at some other time...

pzelasko commented:
Note to self: for the purpose of checking if the total number of frames in batch is lesser than `max_frames`, I'm not taking padding into account atm - should probably change that.

danpovey commented on lhotse/cut.py:
```diff
@@ -37,6 +37,19 @@ class CutUtilsMixin:
+        that exceed the Cut's duration (it means the supervision continued in the original recording
+        after the Cut's ending).
+        """
+        return [s.trim(self.duration) for s in self.supervisions]
```
For ASR tasks, it would only be useful to train with trimmed cuts if we knew that they were trimmed (and on which ends) and could modify the supervision FSA accordingly.

pzelasko replied:
What do you think is the right way to return that information? It would be straightforward to return negative `start_frame` and `end_frame` exceeding the time dimension in the features, but I'm not sure if that is sufficient for K2 to adjust the supervision FSA (or how it would be adjusted). I'm curious what're your thoughts on this.

Alternatively, we can put an assertion in this dataset to ensure that there are no supervisions that need to be trimmed (at least for now).

danpovey replied:
k2 would probably insert optional entry points and exit points.  pair of bool might be enough.
But I think it's better, for now, to just asser that this doesn't happen.  It's not something I've ever done in ASR training (cutting utterance without knowing the alignment).

danpovey commented on lhotse/dataset/speech_recognition.py:
```diff
@@ -60,6 +62,172 @@ def __len__(self) -> int:
         return len(self.cut_ids)
 
 
+class K2SpeechRecognitionIterableDataset(IterableDataset):
```
I think i would be clearer to inherit from `torch.IterableDataset` explicitly, so readers will know it's not a k2 thing.

danpovey commented on lhotse/dataset/speech_recognition.py:
```diff
@@ -60,6 +62,172 @@ def __len__(self) -> int:
+    The 'sequence_idx' field is the index of the Cut used to create the example in the Dataset.
+    """
+
+    def __init__(
```
Would be nice to have some documentation for things like max_frames!

pzelasko commented:
OK, I added documentation and implemented a first "smarter" batching approach. @danpovey please take a look if that works better.

The good news is that it seems to work OK when I iterate through `test-clean-100` and look at some batches (it also passes through the whole set without errors). 

The bad news is that I had to work around some off-by-one errors that are increasingly tough to address at the root of the problem. I think our assumption that `num_frames` and `duration` would have a simple relationship is not true; perhaps even not possible. It's quite difficult to account for float rounding (0.005) in mixed cuts, especially in cases where it consists of 3+ other cuts and there might be offsets that also have a 5 at the 3rd decimal place. I think I'm too tired ATM to come up with a systematic solution so I'm "fixing" it with tensor slicing during collation.

danpovey commented:
Isn't the num-frames also part of the metadata? ... perhaps that could be
used to decide everything?  I mean convert everything to frame counts at
the outset.

On Sat, Nov 7, 2020 at 1:21 PM Piotr Żelasko <notifications@github.com>
wrote:

> OK, I added documentation and implemented a first "smarter" batching
> approach. @danpovey <https://github.com/danpovey> please take a look if
> that works better.
>
> The good news is that it seems to work OK when I iterate through
> test-clean-100 and look at some batches (it also passes through the whole
> set without errors).
>
> The bad news is that I had to work around some off-by-one errors that are
> increasingly tough to address at the root of the problem. I think our
> assumption that num_frames and duration would have a simple relationship
> is not true; perhaps even not possible. It's quite difficult to account for
> float rounding (0.005) in mixed cuts, especially in cases where it consists
> of 3+ other cuts and there might be offsets that also have a 5 at the 3rd
> decimal place. I think I'm too tired ATM to come up with a systematic
> solution so I'm "fixing" it with tensor slicing during collation.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lhotse-speech/lhotse/pull/116#issuecomment-723394184>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZFLOYSQ333VC2U7LDKVPLSOTKNHANCNFSM4TLY3TQQ>
> .
>

pzelasko commented:
It is, but at this moment there are some edge cases where that meta-data is incorrect. I resolved most of them, but mixed cuts that consist of multiple cuts seem to still suffer from that, and I need to have a fresh mind before I look at it again...

pzelasko commented:
Uh, it was simpler than I thought, just needed a fresh look on this. Anyway, I'm setting up a new issue to make randomized tests that check the invariants of the Cut classes - I want to be more confident that the code does what I think it does, even if it seems to work on LibriSpeech without issues now.

@danpovey please check if what the new dataset class is doing makes sense to you.

pzelasko commented:
FYI I ran it on the CLSP grid, loading `train-clean-100` data for 10 epochs with randomization and "packing" cuts together for more efficient batching, this is the benchmark of how long it took to read + collate data (each row represents an epoch, number of iterations == number of batches). The features were stored on NFS, in separate files per array with lilcom compression on. I used 8 DataLoader parallel workers.

```
1435it [01:01, 23.33it/s]
1434it [00:42, 33.66it/s]
1435it [00:37, 38.09it/s]
1434it [00:37, 38.03it/s]
1436it [00:35, 40.07it/s]
1434it [00:33, 42.45it/s]
1434it [00:34, 41.96it/s]
1433it [00:33, 42.89it/s]
1435it [00:32, 44.39it/s]
1435it [00:32, 43.57it/s]
```

The mean time spent waiting to get a batch was 26ms, and the median was 1ms. I expect it to be even less when the consumer thread will actually do some work in between of reading batches.

I also found one more issue with `MixedCut` num_frames that I patched. After that, the whole 10 data loading epochs passed without errors.

It's ready to merge from my side.

qindazhu commented:
+2, it would be great if you can update pip package after merging this PR,  as I want to use `K2SpeechRecognitionDataset` to build recipes (seems it was not exported in current pip package)

danpovey commented on lhotse/cut.py:
```diff
@@ -244,7 +258,7 @@ def recording_id(self) -> str:
     @property
     def end(self) -> Seconds:
-        return round(self.start + self.duration, ndigits=3)
+        return round(self.start + self.duration, ndigits=8)
```
just wondering why we need to do rounding here... is there a rationale?

danpovey commented on lhotse/cut.py:
```diff
@@ -767,12 +801,19 @@ def load_features(self, mixed: bool = True) -> Optional[np.ndarray]:
+            #   with a regular Cut at the end, when the mix offsets are floats with a lot of decimals.
+            #   For now we're duplicating the last frame to match the declared "num_frames" of this cut.
+            if feats.shape[0] - self.num_frames == -1:
+                feats = np.concatenate((feats, feats[-1:, :]), axis=0)
```
Perhaps assert, after all of this, that feats.shape[0] == self.num_frames, so examples that differ by more than one crash early.

pzelasko commented:
Yeah, limited float arithmetic precision. I originally put 3 decimals there
without too much thought, but it was too coarse. With 8 I think it's enough
not to mess up the duration and num_frames/num_samples relationship. If I
don't round it, it will result in num_frames off by one errors in cuts with
durations like 10.444999999998...

Maybe it'd be cleaner to use the Decimal type instead? I think we'll run
into these problems again with high resolution recordings (44.1kHz and
higher fs).

pon., 9 lis 2020, 22:39 użytkownik Daniel Povey <notifications@github.com>
napisał:

> *@danpovey* commented on this pull request.
> ------------------------------
>
> In lhotse/cut.py
> <https://github.com/lhotse-speech/lhotse/pull/116#discussion_r520265997>:
>
> > @@ -244,7 +258,7 @@ def recording_id(self) -> str:
>
>      @property
>      def end(self) -> Seconds:
> -        return round(self.start + self.duration, ndigits=3)
> +        return round(self.start + self.duration, ndigits=8)
>
> just wondering why we need to do rounding here... is there a rationale?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/lhotse-speech/lhotse/pull/116#pullrequestreview-526829335>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADZRKQF3KDZCAAZ7G5BHMEDSPCYV3ANCNFSM4TLY3TQQ>
> .
>

danpovey commented:
Up to you; if it works, it's OK with me.


On Tue, Nov 10, 2020 at 12:00 PM Piotr Żelasko <notifications@github.com>
wrote:

> Yeah, limited float arithmetic precision. I originally put 3 decimals there
> without too much thought, but it was too coarse. With 8 I think it's enough
> not to mess up the duration and num_frames/num_samples relationship. If I
> don't round it, it will result in num_frames off by one errors in cuts with
> durations like 10.444999999998...
>
> Maybe it'd be cleaner to use the Decimal type instead? I think we'll run
> into these problems again with high resolution recordings (44.1kHz and
> higher fs).
>
> pon., 9 lis 2020, 22:39 użytkownik Daniel Povey <notifications@github.com>
> napisał:
>
> > *@danpovey* commented on this pull request.
> > ------------------------------
> >
> > In lhotse/cut.py
> > <https://github.com/lhotse-speech/lhotse/pull/116#discussion_r520265997
> >:
> >
> > > @@ -244,7 +258,7 @@ def recording_id(self) -> str:
> >
> > @property
> > def end(self) -> Seconds:
> > - return round(self.start + self.duration, ndigits=3)
> > + return round(self.start + self.duration, ndigits=8)
> >
> > just wondering why we need to do rounding here... is there a rationale?
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/lhotse-speech/lhotse/pull/116#pullrequestreview-526829335
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/ADZRKQF3KDZCAAZ7G5BHMEDSPCYV3ANCNFSM4TLY3TQQ
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lhotse-speech/lhotse/pull/116#issuecomment-724434659>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZFLO7IYU7AM5GHMRMQQBDSPC3EFANCNFSM4TLY3TQQ>
> .
>

qindazhu commented:
Got below error when I was trying to use `K2SpeechRecognitionIterableDataset` with `Mfcc` feature,
```
>>> dataset = K2SpeechRecognitionIterableDataset(cuts_test, shuffle=False)  #cuts_test using Mfcc featrue extractor.
>>> dloader = DataLoader(dataset, batch_size=None, num_workers=num_workers)
>>> batch = next(iter(dloader))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/linuxbrew/.linuxbrew/Cellar/python@3.8/3.8.6_1/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 363, in __next__
    data = self._next_data()
  File "/home/linuxbrew/.linuxbrew/Cellar/python@3.8/3.8.6_1/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 989, in _next_data
    return self._process_data(data)
  File "/home/linuxbrew/.linuxbrew/Cellar/python@3.8/3.8.6_1/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 1014, in _process_data
    data.reraise()
  File "/home/linuxbrew/.linuxbrew/Cellar/python@3.8/3.8.6_1/lib/python3.8/site-packages/torch/_utils.py", line 395, in reraise
    raise self.exc_type(msg)
ValueError: Caught ValueError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/home/linuxbrew/.linuxbrew/Cellar/python@3.8/3.8.6_1/lib/python3.8/site-packages/torch/utils/data/_utils/worker.py", line 185, in _worker_loop
    data = fetcher.fetch(index)
  File "/home/linuxbrew/.linuxbrew/Cellar/python@3.8/3.8.6_1/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 34, in fetch
    data = next(self.dataset_iter)
  File "/ceph-hw/lhotse/lhotse/dataset/speech_recognition.py", line 206, in __next__
    features = _collate_features(cuts)
  File "/ceph-hw/lhotse/lhotse/dataset/speech_recognition.py", line 280, in _collate_features
    features[idx] = torch.from_numpy(cut.load_features())
  File "/ceph-hw/lhotse/lhotse/cut.py", line 791, in load_features
    mixer = FeatureMixer(
  File "/ceph-hw/lhotse/lhotse/features/mixer.py", line 48, in __init__
    self.reference_energy = feature_extractor.compute_energy(base_feats)
  File "/ceph-hw/lhotse/lhotse/features/base.py", line 100, in compute_energy
    raise ValueError('The feature extractor\'s "compute_energy" is undefined.')
ValueError: The feature extractor's "compute_energy" is undefined.
```
It's OK when using `Fbank()` together with `K2SpeechRecognitionIterableDataset`,Maybe it's better to add some documentation about this to say we can't use Mfcc for `K2SpeechRecognitionIterableDataset` (or there is already one but I don't know)?

danpovey commented:
It might be that you were missing some args or options.   Maybe search in
lhotse code for examples, e.g. git grep Mfcc ?

On Tue, Nov 10, 2020 at 8:32 PM Haowen Qiu <notifications@github.com> wrote:

> Got below error when I was trying to use
> K2SpeechRecognitionIterableDataset with Mfcc feature,
>
> >>> dataset = K2SpeechRecognitionIterableDataset(cuts_test, shuffle=False)  #cuts_test using Mfcc featrue extractor.
> >>> dloader = DataLoader(dataset, batch_size=None, num_workers=num_workers)
> >>> batch = next(iter(dloader))
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/home/linuxbrew/.linuxbrew/Cellar/python@3.8/3.8.6_1/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 363, in __next__
>     data = self._next_data()
>   File "/home/linuxbrew/.linuxbrew/Cellar/python@3.8/3.8.6_1/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 989, in _next_data
>     return self._process_data(data)
>   File "/home/linuxbrew/.linuxbrew/Cellar/python@3.8/3.8.6_1/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 1014, in _process_data
>     data.reraise()
>   File "/home/linuxbrew/.linuxbrew/Cellar/python@3.8/3.8.6_1/lib/python3.8/site-packages/torch/_utils.py", line 395, in reraise
>     raise self.exc_type(msg)
> ValueError: Caught ValueError in DataLoader worker process 0.
> Original Traceback (most recent call last):
>   File "/home/linuxbrew/.linuxbrew/Cellar/python@3.8/3.8.6_1/lib/python3.8/site-packages/torch/utils/data/_utils/worker.py", line 185, in _worker_loop
>     data = fetcher.fetch(index)
>   File "/home/linuxbrew/.linuxbrew/Cellar/python@3.8/3.8.6_1/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 34, in fetch
>     data = next(self.dataset_iter)
>   File "/ceph-hw/lhotse/lhotse/dataset/speech_recognition.py", line 206, in __next__
>     features = _collate_features(cuts)
>   File "/ceph-hw/lhotse/lhotse/dataset/speech_recognition.py", line 280, in _collate_features
>     features[idx] = torch.from_numpy(cut.load_features())
>   File "/ceph-hw/lhotse/lhotse/cut.py", line 791, in load_features
>     mixer = FeatureMixer(
>   File "/ceph-hw/lhotse/lhotse/features/mixer.py", line 48, in __init__
>     self.reference_energy = feature_extractor.compute_energy(base_feats)
>   File "/ceph-hw/lhotse/lhotse/features/base.py", line 100, in compute_energy
>     raise ValueError('The feature extractor\'s "compute_energy" is undefined.')
> ValueError: The feature extractor's "compute_energy" is undefined.
>
> It's OK when using Fbank() together with
> K2SpeechRecognitionIterableDataset,Maybe it's better to add some
> documentation about this to say we can't use Mfcc for
> K2SpeechRecognitionIterableDataset (or there is already one but I don't
> know)?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lhotse-speech/lhotse/pull/116#issuecomment-724672727>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZFLO4C6DJNFOAGWJ35NE3SPEXFJANCNFSM4TLY3TQQ>
> .
>

danpovey commented:
Oh I see, it's missing code, not options.  It should be possible to compute
the energy, by reversing the transform.
However there's probably no harm in using the fbank features.  We can
always decorrelate them by applying the
cosine transform and then batchnorm.


On Tue, Nov 10, 2020 at 8:47 PM Daniel Povey <dpovey@gmail.com> wrote:

> It might be that you were missing some args or options.   Maybe search in
> lhotse code for examples, e.g. git grep Mfcc ?
>
> On Tue, Nov 10, 2020 at 8:32 PM Haowen Qiu <notifications@github.com>
> wrote:
>
>> Got below error when I was trying to use
>> K2SpeechRecognitionIterableDataset with Mfcc feature,
>>
>> >>> dataset = K2SpeechRecognitionIterableDataset(cuts_test, shuffle=False)  #cuts_test using Mfcc featrue extractor.
>> >>> dloader = DataLoader(dataset, batch_size=None, num_workers=num_workers)
>> >>> batch = next(iter(dloader))
>> Traceback (most recent call last):
>>   File "<stdin>", line 1, in <module>
>>   File "/home/linuxbrew/.linuxbrew/Cellar/python@3.8/3.8.6_1/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 363, in __next__
>>     data = self._next_data()
>>   File "/home/linuxbrew/.linuxbrew/Cellar/python@3.8/3.8.6_1/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 989, in _next_data
>>     return self._process_data(data)
>>   File "/home/linuxbrew/.linuxbrew/Cellar/python@3.8/3.8.6_1/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 1014, in _process_data
>>     data.reraise()
>>   File "/home/linuxbrew/.linuxbrew/Cellar/python@3.8/3.8.6_1/lib/python3.8/site-packages/torch/_utils.py", line 395, in reraise
>>     raise self.exc_type(msg)
>> ValueError: Caught ValueError in DataLoader worker process 0.
>> Original Traceback (most recent call last):
>>   File "/home/linuxbrew/.linuxbrew/Cellar/python@3.8/3.8.6_1/lib/python3.8/site-packages/torch/utils/data/_utils/worker.py", line 185, in _worker_loop
>>     data = fetcher.fetch(index)
>>   File "/home/linuxbrew/.linuxbrew/Cellar/python@3.8/3.8.6_1/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 34, in fetch
>>     data = next(self.dataset_iter)
>>   File "/ceph-hw/lhotse/lhotse/dataset/speech_recognition.py", line 206, in __next__
>>     features = _collate_features(cuts)
>>   File "/ceph-hw/lhotse/lhotse/dataset/speech_recognition.py", line 280, in _collate_features
>>     features[idx] = torch.from_numpy(cut.load_features())
>>   File "/ceph-hw/lhotse/lhotse/cut.py", line 791, in load_features
>>     mixer = FeatureMixer(
>>   File "/ceph-hw/lhotse/lhotse/features/mixer.py", line 48, in __init__
>>     self.reference_energy = feature_extractor.compute_energy(base_feats)
>>   File "/ceph-hw/lhotse/lhotse/features/base.py", line 100, in compute_energy
>>     raise ValueError('The feature extractor\'s "compute_energy" is undefined.')
>> ValueError: The feature extractor's "compute_energy" is undefined.
>>
>> It's OK when using Fbank() together with
>> K2SpeechRecognitionIterableDataset,Maybe it's better to add some
>> documentation about this to say we can't use Mfcc for
>> K2SpeechRecognitionIterableDataset (or there is already one but I don't
>> know)?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/lhotse-speech/lhotse/pull/116#issuecomment-724672727>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAZFLO4C6DJNFOAGWJ35NE3SPEXFJANCNFSM4TLY3TQQ>
>> .
>>
>

pzelasko commented:
@qindazhu You'll getting this error because we haven't implemented feature-domain mixing for MFCC features. I think I attempted it, but was getting too big a difference between the features extracted from mixed audio vs. mixing the features extracted from separate audios, and figured out we can work on it later if needed. If you need it, can you open an issue?

pzelasko commented:
Anyway, the least I can do for now is to improve the error message before merging.

qindazhu commented:
Not urgent, we can use `Fbank` for now. Thanks for the reply, I'm just learning to use lhotse.

pzelasko commented:
Re: the num_frames and rounding thing, I am not 100% happy with the current solution but it seems to work for now so I'll address the review remarks and then merge. I'm thinking going forward maybe we should count the offsets and durations in samples rather than seconds, thus avoiding the float precision problems. If you know the sampling rate and frame shift, you can always derive duration and num frames from that info precisely. That'd be a fairly big change so I'd like to know your opinions guys before attempting it.

danpovey commented:
For things like MFCCs that don't have samples (assuming you've already discarded the wave data) I'm not sure if it makes sense to count samples?   Presumably if you appended those at the feature level it would be based on the frame count.  Maybe the problem there, if you want to be super exact about the length, is that an object like that really has 2 durations, one if you consider it a sequence of frames and another if you consider it a sequence of samples.  And of course if you were to resample, the length would change slightly too.

pzelasko commented:
Maybe you're right. I'll let it be for some time. Then I'll add the randomized tests and we'll see what kinds of issues come up.
